stdlib/in_channel.ml

```diff
@@ -63,4 +78,83 @@ let really_input_string ic len =
   | s -> Some s
   | exception End_of_file -> None
 
+(* Read up to [len] bytes into [buf], starting at [ofs]. Return total bytes
+   read. *)
+let read_upto ic buf ofs len =
+  let rec loop ofs len =
+    if len = 0 then ofs
+    else begin
+      let r = Stdlib.input ic buf ofs len in
+      if r = 0 then
+        ofs
+      else
+        loop (ofs + r) (len - r)
+    end
+  in
+  loop ofs len - ofs
+
+(* Return a buffer that has >= [ofs + n] bytes of storage, and
+   coincides with [buf] at indices < [ofs]. *)
+let ensure buf ofs n =
+  let len = Bytes.length buf in
+  if len >= ofs + n then buf
+  else begin
+    let new_len = ref len in
+    while !new_len < ofs + n do
+      new_len := 2 * !new_len + 1
+    done;
+    let new_len = !new_len in
+    let new_len =
+      if new_len <= Sys.max_string_length then
+        new_len
+      else if ofs < Sys.max_string_length then
+        Sys.max_string_length
```

Comment:
`ensure` doesn't respect the contract written in the doc comment in this case. For instance if `ofs = Sys.max_string_length - 10` and `n = 20`, then the returned buffer will only have `Sys.max_string_length` bytes of storage rather than `Sys.max_string_length + 10`.

Looking at the way this function is used inside `let rec loop ...` bellow, this can be an issue as we call `read_upto` right after with `len = chunk_size`. This means that we might end up doing a call to `Stdlib.input` with `ofs + len > Bytes.length buf`, which would raise `Invalid_argument "input"`.

Reply:
Argh, you are right... I pushed a fix for this problem together with an improved docstring for `read_upto` that explains the exact contract for the function. Thanks!

Reply:
Thanks, looks correct to me now!